clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -0,0 +1,366 @@
+      Map<CharSequence, CharSequence> keySchema = metadataResponse.getKeySchema();
+      Map<CharSequence, CharSequence> valueSchemas = metadataResponse.getValueSchemas();
+      // Map<CharSequence, Integer> helixGroupInfo = metadataResponse.getHelixGroupInfo();
+      Map<Integer, List<String>> routingInfo = metadataResponse.getRoutingInfo()
+          .entrySet()
+          .stream()
+          .collect(
+              Collectors.toMap(
+                  e -> Integer.valueOf(e.getKey().toString()),
+                  e -> e.getValue().stream().map(CharSequence::toString).collect(Collectors.toList())));
+
+      if (fetchedVersion != getCurrentStoreVersion()) {
+        // call the DICTIONARY endpoint if needed
+        CompletableFuture<TransportClientResponse> dictionaryFetchFuture = null;
+        if (!versionZstdDictionaryMap.containsKey(fetchedVersion)
+            && compressionStrategy == CompressionStrategy.ZSTD_WITH_DICT.getValue()) {
+          dictionaryFetchFuture = fetchCompressionDictionary(fetchedVersion);
+        }
+
+        // Update partitioner pair map (versionPartitionerMap)
+        Properties params = new Properties();
+        params.putAll(partitionerParams);
+        VenicePartitioner partitioner =
+            PartitionUtils.getVenicePartitioner(partitionerClass, amplificationFactor, new VeniceProperties(params));
+        versionPartitionerMap.put(fetchedVersion, partitioner);
+        versionPartitionCountMap.put(fetchedVersion, partitionCount);
+
+        // Update readyToServeInstanceMap
+        for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+          String key = getVersionPartitionMapKey(fetchedVersion, partitionId);
+          readyToServeInstancesMap.put(key, routingInfo.get(partitionId));
+        }
+
+        // Update schemas
+        SchemaData schemaData = new SchemaData(storeName);
+        for (Map.Entry<CharSequence, CharSequence> entry: keySchema.entrySet()) {
+          schemaData
+              .setKeySchema(new SchemaEntry(Integer.parseInt(entry.getKey().toString()), entry.getValue().toString()));
+        }
+        for (Map.Entry<CharSequence, CharSequence> entry: valueSchemas.entrySet()) {
+          schemaData.addValueSchema(
+              new SchemaEntry(Integer.parseInt(entry.getKey().toString()), entry.getValue().toString()));
+        }
+        schemas.set(schemaData);
+
+        // Wait for dictionary fetch to finish if there is one
+        try {
+          if (dictionaryFetchFuture != null) {
+            dictionaryFetchFuture.get(ZSTD_DICT_FETCH_TIMEOUT, TimeUnit.SECONDS);
+          }
+          currentVersion.set(fetchedVersion);
+          clusterStats.updateCurrentVersion(getCurrentStoreVersion());
+          latestSuperSetValueSchemaId.set(newSuperSetValueSchemaId);
+          updateComplete = true;
+        } catch (ExecutionException | TimeoutException e) {
+          LOGGER.warn(
```

Comment:
Makes sense, I flipped the logic so `updateComplete` starts off as true and is only set to false in the catch clause. This way it will be true if we run a valid refresh but the metadata is unchanged. If the metadata fetch fails, then `clientStats.updateCacheTimestamp` is never run at all